cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/Dag.java

```diff
@@ -293,15 +295,22 @@ void validate() {
 
   /**
    * Return all stages accessible from the starting stages, without going past any node in stopNodes.
+   * Starting stages are not treated as stop nodes, even if they are in the stop nodes set.
    *
    * @param stages the stages to start at
    * @param stopNodes set of nodes to stop traversal on
    * @return all stages accessible from that stage
    */
   public Set<String> accessibleFrom(Set<String> stages, Set<String> stopNodes) {
     Set<String> accessible = new HashSet<>();
+    final Set<String> nonStartingStopNodes = Sets.difference(stopNodes, stages);
     for (String stage : stages) {
-      accessible.addAll(traverseForwards(stage, accessible, stopNodes));
+      traverseForwards(stage, accessible, new Predicate<String>() {
+        @Override
+        public boolean apply(@Nullable String input) {
```

Comment:
seems like same predicate is used for `traverseForwards` and `traverseBackwards`. Probably better to make it class member?